v3.0.0 Pre-Release

BryceVichiOJ opened the pull request:
## [3.0.0]
### Added
- Added a base driver class mechaspin::parakeet::Driver which holds common functionality between sensor drivers
- Added a driver class to support the Parakeet ProE (mechaspin::parakeet::ProE::Driver)

### Modified
- Updated SimpleExample to support the Parakeet ProE sensor
- [BREAKING] Replaced mechaspin::parakeet::Driver with mechaspin::parakeet::Pro::Driver
- [BREAKING] Replaced PointPolar.getRange() with PointPolar.getRange_mm()
- [BREAKING] Replaced PointPolar.getAngleInDegrees() with PointPolar.getAngle_deg()

BryceVichiOJ committed:
PRKT-89 Add Base Class for Driver

BryceVichiOJ committed:
Merge pull request #22 from MechaSpin/PRKT-89_Add_Base_Class_for_Driver

PRKT-89 Add Base Class for Driver

BryceVichiOJ committed:
PRKT-87 Add Ethernet Class to SDK

BryceVichiOJ committed:
Merge pull request #23 from MechaSpin/PRKT-87_Add_Ethernet_class_to_SDK

PRKT-87 Add Ethernet Class to SDK

BryceVichiOJ committed:
PRKT-103 Checksum checking

BryceVichiOJ committed:
Merge pull request #24 from MechaSpin/PRKT-103_Checksum_checking

PRKT-103 Checksum checking

BryceVichiOJ committed:
PRKT-96 Including Pro Driver before ProE Driver

BryceVichiOJ committed:
Merge pull request #25 from MechaSpin/PRKT-96_Including_Pro_Driver_before_ProE_Driver

PRKT-96 Including Pro Driver before ProE Driver

BryceVichiOJ committed:
PRKT-105 Update Application Defaults

BryceVichiOJ committed:
Merge pull request #26 from MechaSpin/PRKT-105_Update_Application_Defaults

PRKT-105 Update Application Defaults

BryceVichiOJ committed:
PRKT-93 Remove Deprecated from v2.0.0

BryceVichiOJ committed:
Merge pull request #27 from MechaSpin/PRKT-93_Remove_Deprecated

PRKT-93 Remove Deprecated from v2.0.0

BryceVichiOJ committed:
PRKT-102 Initial Connection Failure not descriptive

BryceVichiOJ committed:
Merge pull request #28 from MechaSpin/PRKT-102_Initial_Connection_Failure

PRKT-102 Initial Connection Failure not descriptive

BryceVichiOJ committed:
PRKT-98 Test Serial Driver

BryceVichiOJ committed:
Merge pull request #29 from MechaSpin/PRKT-98_Test_Serial_Driver

PRKT-98 Test Serial Driver

BryceVichiOJ committed:
Fix Linux Compile Issues

BryceVichiOJ committed:
Merge pull request #31 from MechaSpin/Fix_Linux_Compile_Issues

Fix Linux Compile Issues

BryceVichiOJ committed:
PRKT-95 Use Timestamp From Sensor

BryceVichiOJ committed:
Merge pull request #32 from MechaSpin/PRKT-95_Use_Timestamp_From_Sensor

PRKT-95 Use Timestamp From Sensor

BryceVichiOJ committed:
PRKT-106 Provide Doxygen Comments

BryceVichiOJ committed:
Merge pull request #33 from MechaSpin/PRKT-106_Provide_Doxygen_Comments

PRKT-106 Provide Doxygen Comments

BryceVichiOJ committed:
PRKT-104 Refactor EthernetPort to UdpSocket

BryceVichiOJ committed:
Merge pull request #34 from MechaSpin/PRKT-104_Refactor_EthernetPort_to_UdpSocket

PRKT-104 Refactor EthernetPort to UdpSocket

BryceVichiOJ committed:
PRKT-104 Refactor EthernetPort Follow Up

Missed a couple items which were part of PRKT-104. Fixed them here

BryceVichiOJ committed:
Merge pull request #35 from MechaSpin/PRKT-104_Refactor_EthernetPort_Follow_Up

PRKT-104 Refactor EthernetPort Follow Up

BryceVichiOJ committed:
PRKT-99 Update Documentation

BryceVichiOJ committed:
Merge pull request #36 from MechaSpin/PRKT-99_Update_Documentation

PRKT-99 Update Documentation

BryceVichiOJ committed:
PRKT-107 Finalize Changelog / CMake Version

BryceVichiOJ committed:
Merge pull request #37 from MechaSpin/PRKT-107_Finalize_Changelog_CMake_Version

PRKT-107 Finalize Changelog / CMake Version

BryceVichiOJ committed:
PRKT-108 Add Support for modifying IP Address

BryceVichiOJ committed:
Merge pull request #39 from MechaSpin/PRKT-108_Add_Support_for_modifying_IP_Address

PRKT-108 Add Support for modifying IP Address

BryceVichiOJ commented:
Working on PRKT-110. Let's hold off on merging until that is completed.

nichmj01 commented on include/parakeet/Pro/Driver.h:
```diff
@@ -64,23 +56,16 @@ class Driver
-        PARAKEET_DEPRECATED(bool connect(const std::string& comPort, BaudRate baudRate, bool intensity, ScanningFrequency scanningFrequency_Hz, bool dataSmoothing, bool dragPointRemoval));
-
         /// \brief Start the Driver's processing thread
         void start();
```
Let's mark these methods with override: start, stop, close. This also applies to the ProE::Driver.

nichmj01 commented on src/parakeet/Driver.cpp:
```diff
@@ -0,0 +1,124 @@
+
+	void Driver::close()
+	{
+        bool isc = isConnected();
```
What's this variable for? Doesn't seem like it's used. Actually, none of the variables in here seems like they are used?

nichmj01 commented on src/parakeet/Driver.cpp:
```diff
@@ -0,0 +1,124 @@
+        }
+	}
+
+    bool Driver::isUpdateThreadRunning()
```
Just isRunning() might be a better name? Having UpdateThread here seems to add detail that's not necessary or useful?

nichmj01 commented on src/parakeet/Driver.cpp:
```diff
@@ -0,0 +1,124 @@
+        updateThreadCallbackFunction = callback;
+    }
+    
+    void Driver::onScanDataReceived(ScanData* scanData)
```
It would be better to use a const ScanData& for the parameter. This prevents this method from accidentally modifying the scan data , which it shouldn't be doing, and maintains the responsibility lifetime management to the caller.

You transfer ownership here from the caller to this method which in unnecessary here since this method only really cares to use the ScanData, not what happens with it after it's done using it.

If you really need to transfer ownership I would suggest the use of the unique_ptr to help indicate that ownership has been transferred. As it is, it's ambiguous as to who owns the pointer after onScanDataReceived is called which could easily lead to memory leaks. 

See sections F and R here: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-conventional

nichmj01 commented on src/parakeet/Pro/Driver.cpp:
```diff
@@ -432,12 +337,12 @@ namespace Pro
                 }
 
                 ScanData* data = new ScanData;
```
You currently delete the ScanData* object inside the onScanDataReceived method so it seems like you don't really need a pointer here if you change the onScanDataReceived as I suggest and pass a const ScanData& instead of transferring ownership.

nichmj01 commented on src/parakeet/Driver.cpp:
```diff
@@ -0,0 +1,124 @@
+		}
+	}
+
+	void Driver::throwExceptionIfNotConnected()
```
Consider renaming this to something assertIsConnected() just to shorten up the method name. I feel like the word assery as pretty clear meaning; this thing MUST be true otherwise there's an issue.

nichmj01 commented on src/parakeet/Pro/Driver.cpp:
```diff
@@ -70,12 +59,21 @@ namespace Pro
+    {
+        this->registerUpdateThreadCallback(std::bind(&Driver::serialUpdateThreadFunction, this));
+
+        serialPortDataBuffer = new unsigned char[SERIAL_MESSAGE_DATA_BUFFER_SIZE];
```
Seems like there's no reason to use a pointer here. It looks like we have the same size buffer for the lifetime of this owning object.

nichmj01 commented on src/parakeet/ProE/Driver.cpp:
```diff
@@ -0,0 +1,426 @@
+
+        ethernetPortDataBufferLength += charsRead;
+
+        unsigned int nl = parser.parseSensorBuffer(ethernetPortDataBufferLength, ethernetPortDataBuffer);
```
What is nl here? length of data parsed?

BryceVichiOJ replied:
Yeah, its the amount of data parsed out of the buffer. We could have a message and a half in the buffer. We want to parse out only the first message and leave the other half in the buffer for later reading. I'll rename it to bytesParsed.

nichmj01 commented on src/parakeet/ProE/internal/Parser.cpp:
```diff
@@ -0,0 +1,348 @@
+            return;
+        }
+
+        CompleteLidarMessage* lidarMessage = new CompleteLidarMessage();
```
Do we need a pointer here? Seems like we can use a non-pointer instance here and use the same approach as I suggested for the ScanData and the onScanDataReceived callback.

nichmj01 commented on include/parakeet/ProE/internal/Parser.h:
```diff
@@ -0,0 +1,139 @@
+
+		Parser(std::function<void(CompleteLidarMessage*)> onCompleteLidarMessageCallback);
+
+		int parseSensorBuffer(int length, unsigned char* buf);
```
I'd probably just call this method parse. For more context I'd probably change the class name to MessageParser or CommandParser or something. For the UdpSocket class we have the char* parameter first and the length second, where here we have the length first and the char* second. I think it'd be good for it to be consistent. Possibly speaks to it being useful to hav a simple Data or Buffer type struct that groups the buffer and length together. Seems like that could be used in multiple places.

nichmj01 commented on include/parakeet/UdpSocket.h:
```diff
@@ -0,0 +1,72 @@
+	/// \param[in] currentLength - The number of bytes used in the current buffer
+	/// \param[in] bufferSize - The maximum size of the buffer
+	/// \returns The number of bytes read from the stream
+	int read(unsigned char* buffer, int currentLength, int bufferSize);
```
For read we have an unsigned char* and for write we have a (signed) char*, is there a reason for that? I can't see where it would make a difference as at this point I just want to send the bytes on the wire, I'm not interpreting it.

nichmj01 commented on include/parakeet/ProE/internal/Parser.h:
```diff
@@ -0,0 +1,139 @@
+		void reset();
+
+	private:
+		struct BufferData
```
Yeah, I would probably move this out to be a public standalone struct (likely internal to start) and pass around the struct vs the char* and length up until the end.

BryceVichiOJ replied:
Agree this is the way forward, my question is where it should be located. Keeping it within UdpSocket / Parser does not make sense as they would need to include each other. Should I make a separate file which has these structs? If so, what should it be called? IPv4Structs?

BryceVichiOJ committed:
PRKT-110 Create Utility Application to modify IPv4 settings of ProE

BryceVichiOJ committed:
Merge pull request #40 from MechaSpin/PRKT-110_Create_Utility_Application_to_modify_sensor_settings

PRKT-110 Create Utility Application to modify IPv4 settings of ProE

nichmj01 commented on include/parakeet/UdpSocket.h:
```diff
@@ -0,0 +1,72 @@
+	/// \returns The current connection state
+	bool isConnected();
+private:
+	struct EthernetConnection
```
Not sure this struct is necessary? Doesn't seem to add value over having them as direct member fields.

nichmj01 commented on CHANGELOG.md:
```diff
@@ -4,6 +4,18 @@ All notable changes to this project will be documented in this file.
+
+### Modified
+- Updated SimpleExample to support the Parakeet ProE sensor
+- [BREAKING] Replaced mechaspin::parakeet::Driver with mechaspin::parakeet::Pro::Driver
```
I'd probably have these lines under a Removed secton vs the Modified section. 

Looking back at the v2.0.0, I'd also likely have put lines under Added that just said something like: "Added PointPolar.getRange_mm()" in addition to the updates in the Deprecated section.

nichmj01 commented on CHANGELOG.md:
```diff
@@ -4,6 +4,18 @@ All notable changes to this project will be documented in this file.
+- Added an example application to set the IPv4 settings of a Parakeet-ProE sensor
+
+### Modified
+- Updated SimpleExample to support the Parakeet ProE sensor
```
I kinda like having the changes to the SimpleApplication listed here and having just one CHANGELOG, maybe we should indicate here the version of the SimpleApplication? Something like:

Updated SimpleExample to support the Parakeet ProE sensor (v2.0.0)?

nichmj01 reviewed (commented):
I'd consider renaming the ProEIPv4SettingsExample application to something like ConfigureProEIPv4Settings. I can see the argument that it's an example but it feels more like a utility to me. I'd also consider moving the code to a utils folder vs examples.

BryceVichiOJ commented:
> I'd consider renaming the ProEIPv4SettingsExample application to something like ConfigureProEIPv4Settings. I can see the argument that it's an example but it feels more like a utility to me. I'd also consider moving the code to a utils folder vs examples.

@nichmj01  I thought about making an apps folder and moving the utility app into there. For me, apps are a bit more fleshed out than this application. Incorrect IP addresses can be input without errors being thrown. Hence, I decided it was more of an example of how it could be done. Perhaps its best to move it to an apps folder and create YT issues to come back and add invalid input checks.

BryceVichiOJ committed:
PRKT-112 PR #38 Fixes

BryceVichiOJ committed:
Merge pull request #41 from MechaSpin/PRKT-112_PR#38_Fixes

PRKT-112 PR #38 Fixes

BryceVichiOJ committed:
Linux specific missing include

BryceVichiOJ committed:
Merge pull request #42 from MechaSpin/linux_bug_fix

Linux-specific missing include

nichmj01 commented on docs/Building and Running SimpleExample.md:
````diff
@@ -50,10 +52,16 @@ make
+Parakeet ProE:
 ```
-{PARAKEET_ROOT}/examples/SimpleExample/build/SimpleExample.exe COMPORT BAUDRATE (ie: "/dev/ttyUSB0 500000")
+{PARAKEET_ROOT}/examples/SimpleExample/build/SimpleExample.exe ProE destinationIPAddress destinationPort sourcePort (ie: "ProE 192.168.158.98")
````
Your example input doesn't have a port?

nichmj01 commented on examples/SimpleExample/main.cpp:
```diff
@@ -4,10 +4,11 @@
 #include <parakeet/util.h>
 
-const char versionNumber[] = "1.1.0";
+const char versionNumber[] = "2.0.0";
```
As a comment for the future, I think it would be possible to use a define here which would get defined by cmake based on the value in the CMakeLists.txt. That would eliminate the need to update the version in multiple places.

BryceVichiOJ replied:
Looks like this is done via a config file in CMake. I'll make a YT issue for this.

nichmj01 commented on include/parakeet/Driver.h:
```diff
@@ -2,17 +2,135 @@
 namespace parakeet
 {
-    PARAKEET_DEPRECATED(typedef Pro::Driver Driver);
+class Driver
```
We need to make the destructor virtual. I don't see a specifically called out destructor here (which I believe we said means it will be non-virtual) and did notice that the derived classes (ie Pro::Driver & ProE::Driver) have non-virtual destructors.

[] Add virtual destructor
[] Make destructor on derived classes virtual

nichmj01 commented on src/parakeet/Pro/Driver.cpp:
```diff
@@ -260,76 +217,56 @@ namespace Pro
-    {
-        unsigned char* line = new unsigned char[MESSAGE_DATA_BUFFER_SIZE];
-        int len = 0;
+        int charsRead = serialPort.read(serialPortDataBuffer, serialPortDataBufferLength, SERIAL_MESSAGE_DATA_BUFFER_SIZE);
```
Seems like we could be using the newly added BufferData struct in a couple places in here. Let's add a YT issue to review this code and see if it makes sense.

BryceVichiOJ replied:
YT issue created.

nichmj01 commented on src/parakeet/UdpSocket.cpp:
```diff
@@ -0,0 +1,180 @@
+		}
+	}
+
+	int UdpSocket::read(const mechaspin::parakeet::internal::BufferData& bufferData, int bufferMaxSize)
```
We probably want to also pass the source InetAddress back to the user as well. Our driver code currently doesn't care but as a stand-alone class, it would be be necessary.

BryceVichiOJ replied:
YT issue created.

nichmj01 reviewed (changes requested):
Only the destructor changes are needed at this point I think.

BryceVichiOJ committed:
Additional PR Fixes

- Added virtual deconstructor for Driver classes
- Added missing parameters to documentation

BryceVichiOJ committed:
Merge pull request #43 from MechaSpin/more_pr_fixes

Additional PR Fixes
